### PR TITLE
#938 - implemented MkMilestone

### PR DESCRIPTION
**Pull request by RomanKisilenko**
Please review pull request.

### CODE REVIEW

**Comment by dmarkov** [comment]
Let me find a reviewer for this pull request, thanks for submitting it

**Comment by dmarkov** [comment]
@krzyk this pull request is for you, please review

**Comment by krzyk** [comment]
@RomanKisilenko travis build fails, have you tried running `mvn clean install -Pqulice`?

**Comment by RomanKisilenko** [comment]
@krzyk Style violations are fixed now, please take a look.

**Comment by RomanKisilenko** [comment]
I have a few observations. 
1. Is there any configuration guide for a popular IDE used by our team? Code formatting style for IDE like IntelliJ Idea would be very helpful.
2. It would be helpful if travis will send an email to developer who have made the change on failed build as it takes time to manually check build results.
3. I encountered an interesting situation related to code style. One rule says that line length should not exceed 80 characters, while other disallow string literal concatenation. How in this situation can I declare a string constant which is more than 80 characters long?

**Comment by krzyk** [comment]
@RomanKisilenko 
1. I don't know really I tuned my idea manually to follow the rules
2. I'm not expert on travis, so I can't help with that
3. You have to use e.g. String.format() or StringUtils.join or Joiner() from guava, and join a string on some separator. If this doesn't help (you have a string that doesn't have separators) you can add exception from checktyle using: `//@checkstyle LineLength (1 line)` and the following line can be longer

**Comment by krzyk** [comment]
@RomanKisilenko You are using indentation of 8 spaces (or a tab), while we use 4 spaces in the project, I don't know why qulice doesn't complain about that.

**Comment by RomanKisilenko** [comment]
@krzyk Fixed indentations, please check.

**Comment by RomanKisilenko** [comment]
@yegor256 I think first two questions I asked most likely are intended for someone who are in charge of project. Here they are:
1. Is there any configuration guide for a popular IDE used by our team? Code formatting style for IDE like IntelliJ Idea would be very helpful.
2. It would be helpful if travis/appveyor will send an email to developer who have made the change on failed build as it takes time to manually check build results.

**Comment by yegor256** [comment]
@RomanKisilenko this is my settings.jar for IntelliJ IDEA: http://img.teamed.io/settings.jar

I'm not sure this feature exists in Travis, for pull requests. Do you know how this can be done?

**Comment by RomanKisilenko** [comment]
@yegor256 Thank you. According to this page http://docs.travis-ci.com/user/notifications/ notifications should be doable. Should I create an issue for this change?

File: src/test/java/com/jcabi/github/mock/MkMilestoneTest.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    private static final String LOGIN = "login";
+
+    /**
+     * Expected xpath for milestone storage.
+     */
+    private static final String EXPECTED_XPATH =
+        "/github/repos/repo[@coords='user/repo']/milestones[number='1']";
+
+    /**
+     * Tests that returned repo has the same coordinates as milestone.
```
Comment (krzyk) [inline src/test/java/com/jcabi/github/mock/MkMilestoneTest.java]: Please see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names regarding description and name of test methods.

File: src/test/java/com/jcabi/github/mock/MkMilestoneTest.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        final Repo repo = milestone.repo();
+        MatcherAssert.assertThat(
+            repo.coordinates(),
+            Matchers.equalTo(coordinates)
+        );
+    }
+
+    /**
+     * Tests that xpath is formatted properly.
+     *
```
Comment (krzyk) [inline src/test/java/com/jcabi/github/mock/MkMilestoneTest.java]: Additional space between Description and prams/throws is redundant

**Comment by krzyk** [comment]
@RomanKisilenko Additional comments above

**Comment by RomanKisilenko** [comment]
@krzyk Changes made, please check.

**Comment by krzyk** [comment]
@RomanKisilenko There is still `compareTo` method that throws `UnsupportedOperationException`

**Comment by RomanKisilenko** [comment]
@krzyk Fixed, please check now.

**Comment by krzyk** [comment]
@rultor merge pls

File: src/main/java/com/jcabi/github/mock/MkMilestone.java [hunk 3]
```diff
@@ -142,4 +151,17 @@ public int number() {
         return this.code;
     }
 
+    /**
+     * XPath of this element in XML tree.
+     * @return XPath
+     */
+    @NotNull(message = "Xpath is never NULL")
+    public String xpath() {
```
Comment (yegor256) [inline src/main/java/com/jcabi/github/mock/MkMilestone.java]: I don't think it's a good idea to make it public. What for?

**Comment by yegor256** [comment]
@RomanKisilenko I didn't get why `MkMilestone#xpath()` is public?

**Comment by RomanKisilenko** [comment]
@yegor256 Actually, it was done to make the method testable as package private and protected are not allowed by checkstyle & PMD.

**Comment by yegor256** [comment]
@RomanKisilenko it's a very bad idea - to make classes more testable by exposing their internals, don't do this

**Comment by RomanKisilenko** [comment]
@yegor256 So, I guess a preferred solution is to remove the test rather than to make the method package private and suppress checktyle violation, correct?

**Comment by yegor256** [comment]
@RomanKisilenko any option is better than what we have now. Exposing class internals to make it testable is even worse than mocking, see http://www.yegor256.com/2014/09/23/built-in-fake-objects.html

**Comment by RomanKisilenko** [comment]
@yegor256 The issue is fixed.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by yegor256** [comment]
@rultor merge it

**Comment by dmarkov** [comment]
@krzyk Thanks a lot, I just topped your account for **18 mins**, transaction ID `50517727`

+18 to your rating, your total score is [+1988](http://www.netbout.com/b/35035?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy

**Pull request merged**